pkg/controller/master/virtualmachine/vm_controller.go

```diff
@@ -367,3 +379,42 @@ func (h *VMController) ManageOwnerOfPVCs(_ string, vm *kubevirtv1.VirtualMachine
 
 	return h.OnVMRemove(vm)
 }
+
+func (h *VMController) SetHaltIfInsufficientResourceQuota(_ string, vm *kubevirtv1.VirtualMachine) (*kubevirtv1.VirtualMachine, error) {
+	if vm == nil || vm.Spec.Template == nil {
+		return vm, nil
+	}
+	if exist, err := h.vmrCalculator.ExistsVMPod(vm.Namespace, vm.Name); err != nil || exist {
+		return vm, err
+	}
+
+	if strategy, err := vm.RunStrategy(); err != nil || strategy == kubevirtv1.RunStrategyHalted {
+		return vm, err
+	}
+
+	if err := h.vmrCalculator.CheckVMLimits(vm); err == nil {
```

Comment:
There is a duplication of code here. `calculator.CheckVMLimits` runs a lot of the same logic that is being executed in this method.